FLUX-190 - vl-select-next - lifecycle fix, FLUX-77 - choices.js@11.1, FLUX-81 - setFormData(), FLUX-121 - vl-select-next - declarative options

Goldflow opened the pull request:
`fix: FLUX-190 - vl-select-rich-next - component lifecycle verbeterd`
`fix: FLUX-77 - vl-select-rich-next - choices.js bijgewerkt tot v11.1`
`fix: FLUX-81 - vl-form-next - introductie setFormData() functie om form data in te stellen`
`fix: FLUX-121 - vl-select-next - declaratieve options toegevoegd`
`fix: vl-map-search - stories werken opnieuw`

[FLUX-190](https://jira.omgeving.vlaanderen.be/jira/browse/FLUX-190)
[FLUX-77](https://jira.omgeving.vlaanderen.be/jira/browse/FLUX-77)
[FLUX-81](https://jira.omgeving.vlaanderen.be/jira/browse/FLUX-81)
[FLUX-121](https://jira.omgeving.vlaanderen.be/jira/browse/FLUX-121)

[Bamboo](https://www.milieuinfo.be/bamboo/browse/FLUX-CFWC90)

Goldflow committed:
fix: FLUX-77 - vl-select-rich-next - choices.js bijgewerkt tot v11.1

(cherry picked from commit 6d470dd81a5fce09ce5206364d298351c47b8a10)

Goldflow committed:
fix: FLUX-81 - vl-form-next - introductie setFormData() functie om form data in te stellen

(cherry picked from commit 91b9caaaf1b78847765269a8a72d319995b7f205)

k-o-e-n commented on libs/form/src/next/select-rich/vl-select-rich.component.ts:
```diff
@@ -57,7 +57,17 @@ export class VlSelectRichComponent extends FormControl {
+                hasChanged: (value, oldValue) => {
+                    if (value instanceof Array && oldValue instanceof Array) {
+                        // We vergelijken de arrays
+                        return JSON.stringify([...value]) !== JSON.stringify([...oldValue]);
```
De spread operator is hier niet nodig. JSON.stringify verandert de oorspronkelijke arrays niet

k-o-e-n replied:
Kan hier ook gebruik gemaakt worden van de `compareValue` functie?

Goldflow replied:
Done

k-o-e-n commented on libs/form/src/next/select-rich/vl-select-rich.component.ts:
```diff
@@ -70,18 +80,20 @@ export class VlSelectRichComponent extends FormControl {
+    static compareValue = (value: unknown, oldValue: unknown) => {
+        if (value instanceof FormData && oldValue instanceof FormData) {
+            // We vergelijken de letterlijke inhoud van de entries van dit FormData object, omdat default FormData vergelijking niet voldoet
+            return JSON.stringify([...value.entries()]) !== JSON.stringify([...oldValue.entries()]);
```
De spread operator is hier ook niet nodig

Goldflow replied:
Aangepast

k-o-e-n replied:
Sorry, dit was wel nodig om van `.entries()` een array te maken (zie mijn volgende comment)

k-o-e-n commented on libs/form/src/next/select-rich/vl-select-rich.component.ts:
```diff
@@ -59,40 +67,60 @@ export class VlSelectRichComponent extends FormControl {
+        const valueList = value instanceof FormData ? value.entries() : value;
+        const oldValueList = oldValue instanceof FormData ? oldValue.entries() : oldValue;
+        if (valueList instanceof Array && oldValueList instanceof Array) {
+            return structuredClone(valueList) !== structuredClone(oldValueList);
```
Oei, ik zie dat mijn vorige comment over het verwijderen van de spread operator bij de `FormData` fout was. De spread is wel nodig om van `.entries()` een array te maken, die je dan met `JSON.stringify()` kan checken.

Dus het zou moeten worden:
```
const valueList = value instanceof FormData ? [...value.entries()] : value;
const oldValueList = oldValue instanceof FormData ? [...oldValue.entries()] : oldValue;
```

Met `structuredClone` vergelijken kan niet, want deze gaan nooit gelijk zijn. De vergelijking checkt de objecten maar niet hun inhoud. De objecten zijn nooit gelijk aan elkaar, want ze zijn net aangemaakt door `structuredClone`.

Het zou dan dit moeten worden:
```
return JSON.stringify(valueList) !== JSON.stringify(oldValueList);
```

We zouden ook kunnen overwegen om `isEqual` van lodash-es te gebruiken:
```
import isEqual from 'lodash-es/isEqual';

return isEqual(valueList, oldValueList);
```
Dit doet een deep equality check en houdt geen rekening met de volgorde van de items.

Goldflow replied:
Klopt, heb ik even te vlug aangepast en het klopt ook ivm `structuredClone` > aangepast
Voor de `v1` branch zou ik geen libraries meer toevoegen maar te overwegen voor `v2` eventueel

Goldflow committed:
fix: FLUX-121 - vl-select-next - declaratieve options toegevoegd

Goldflow committed:
fix: FLUX-190 - vl-select-rich-next - component lifecycle verbeterd

Goldflow committed:
fix: vl-map-search - stories werken opnieuw

Door de map-search & bijhorende map een aparte id te geven, werkt de 2e story opnieuw.

kspeltix commented:
:tada: This PR is included in version 1.49.5 :tada:

The release is available on [GitHub release](https://github.com/milieuinfo/flux-web-components/releases/tag/v1.49.5)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
